source/beamline/BioXAS/BioXASCarbonFilterFarmActuatorFilterControl.cpp

```diff
@@ -0,0 +1,298 @@
+}
+
+int BioXASCarbonFilterFarmActuatorFilterControl::indexOf(double filter) const
+{
+	return filters_.indexOf(filter);
+}
+
+int BioXASCarbonFilterFarmActuatorFilterControl::indexOf(const QString filterString) const
+{
+	return enumNames().indexOf(filterString);
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::setCurrentWindow(BioXASCarbonFilterFarmActuatorWindowControl *newControl)
+{
+	if (currentWindow_ != newControl) {
+
+		if (currentWindow_)
+			removeChildControl(currentWindow_);
+
+		currentWindow_ = newControl;
+
+		if (currentWindow_) {
+			addChildControl(currentWindow_);
+
+			connect( currentWindow_, SIGNAL(windowsChanged()), this, SLOT(updateFilters()) );
+		}
+
+		updateStates();
+
+		emit currentWindowChanged(currentWindow_);
+	}
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::setFilterWindowPreference(double filter, BioXASCarbonFilterFarmWindowOption *window)
+{
+	if (filters_.contains(filter) && window && window->isValid() && currentWindow_->indexOf(window) != -1 && double(window->filter()) == filter)
+		filterWindowPreferenceMap_.insert(filter, window);
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::clearFilterWindowPreference(double filter)
+{
+	if (filterWindowPreferenceMap_.keys().contains(filter))
+		filterWindowPreferenceMap_.remove(filter);
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::clearFilterWindowPreference(BioXASCarbonFilterFarmWindowOption *window)
+{
+	if (filterWindowPreferenceMap_.values().contains(window)) {
+		QList<double> filters = filterWindowPreferenceMap_.keys(window);
+
+		foreach (double filter, filters)
+			filterWindowPreferenceMap_.remove(filter);
+	}
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::clearFilterWindowPreferences()
+{
+	filterWindowPreferenceMap_.clear();
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::updateStates()
+{
+	updateConnected();
+	updateFilters();
+	updateEnumStates();
+	updateMaximumValue();
+	updateValue();
+	updateMoving();
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::updateConnected()
+{
+	bool connected = ( currentWindow_ && currentWindow_->isConnected() );
+	setConnected(connected);
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::updateValue()
+{
+	double newValue = enumNames().indexOf("Unknown");
+
+	if (isConnected()) {
+		BioXASCarbonFilterFarmWindowOption *currentWindow = currentWindow_->windowAt(int(currentWindow_->value()));
+
+		if (currentWindow && currentWindow->isValid()) {
+			double currentFilter = currentWindow->filter();
+			int currentIndex = indexOf(currentFilter);
+
+			if (currentIndex >= 0)
+				newValue = currentIndex;
+		}
+	}
+
+	setValue(newValue);
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::updateMoving()
+{
+	bool isMoving = false;
+
+	if (isConnected())
+		isMoving = (currentWindow_->isMoving());
+
+	setIsMoving(isMoving);
+}
+
+void BioXASCarbonFilterFarmActuatorFilterControl::updateMaximumValue()
+{
+	setMaximumValue(enumNames().count() - 1);
```

Comment:
For this control, is it implied that the maximum value is just the last in the list?  Could this be different somehow or would this be something you'd try and enforce in the implementations?